text/0000-reads-clauses-on-methods.md

```diff
@@ -0,0 +1,216 @@
+{
+  ...
+}
+```
+
+Suppose you want to add some caching as we implement this, since the results of one call can be reused by others. We need to have clients pass in the cache since Dafny offers no way to store it as global state (which is frankly a good thing). Our first attempt leads to a verification error, however, which saves us from accidentally introducing a race condition!
+
+```dafny
+
+class FibonacciMemoizer {
+  var cache: map<nat, nat>
+  constructor() reads {} {
+    cache := map[];
+  }
+  method Get(n: nat) returns (r: nat)
+    reads this
+    modifies this
+  {
+    if n <= 2 {
+      return 1;
+    }
+    if n in cache {
+      r := cache[n];
+    } else {
+      var oneBefore := Get(n - 1);
+      var twoBefore := Get(n - 2);
+      r := oneBefore + twoBefore;
+      // This line introduces a race condition in multi-threaded target languages.
+      // Updating two different entries concurrently can lead to one of them being dropped!
+      cache := cache[n := r];
+    }
+  }
+}
+
+method {:extern} Fibonacci(n: nat, memoizer: FibonacciMemoizer) returns (r: nat)
+  reads {}
+{
+  r := memoizer.Get(n);   // Error: insufficient reads clause to invoke method
+                          // Error: call may violate context's modifies clause
+}
+```
+
+Note that this technique is most effective when the person providing the specification is different from the person implementing it. It ensures that the implementation cannot interact with shared state by accident, especially deep within shared libraries that are perfectly safe to use in a single-threaded context. The wiki page on [Modelling External State Correctly](https://github.com/dafny-lang/dafny/wiki/Modeling-External-State-Correctly) would be updated to recommend that methods exposed to concurrent native code should include `reads {}` and omit any `modifies` clauses.
+
+The nature of frame specifications in Dafny makes `reads {}` less restrictive than it might appear. First of all, it only specifies that a method reads no *mutable* fields. `const` declarations on classes are still fair game. It also only forbids reading *existing* mutable state. It is legal to allocate new mutable objects and modify their fields, either inline or by calling other methods. This means that creating and sharing a cache between multiple calculations within a single externally-exposed method is permitted:
+
+```dafny
+method {:extern} FirstNFibonaccis(n: nat) returns (rs: seq<nat>)
+  reads {}
+{
+  var memoizer := new FibonacciMemoizer();
+  rs := [];
+  for i := 1 to n + 1 {
+    var r := memoizer.Get(i);
+    rs := rs + [r];
+  }
+}
+```
+
+When sharing state externally really is necessary, it must be implemented externally in native code, where native facilities for synchronizing access to mutable state can be used. Therefore the solution is to push the caching logic into the target language instead:
+
+```dafny
+datatype Option<T> = Some(value: T) | None
```

Comment:
I suggest moving this type definition down below the `ExternalMutableMap`. I found myself reading it and the sentence above it several times, trying to figure out their connection, until I realized the sentence above has nothing to do with `datatype`s and that this `datatype` is introduced only to support the example.